// next-sitemap/dist/cjs/url/create-url-set/__tests__/create-url-set.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const config_1 = require("../../../fixtures/config");
const manifest_1 = require("../../../fixtures/manifest");
describe('createUrlSet', () => {
    test('without exclusion', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(config_1.sampleConfig, manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-0',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-1',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-2',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-3',
                alternateRefs: [],
            },
        ]);
    }));
    test('with exclusion', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { exclude: ['/', '/page-0', '/page-2'] }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-1',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-3',
                alternateRefs: [],
            },
        ]);
    }));
    test('with wildcard exclusion', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { exclude: ['/page*'] }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com',
                alternateRefs: [],
            },
        ]);
    }));
    test('without trailing slash', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { trailingSlash: false }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-0',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-1',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-2',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-3',
                alternateRefs: [],
            },
        ]);
    }));
    test('with trailing slash', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { trailingSlash: true }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-0/',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-1/',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-2/',
                alternateRefs: [],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-3/',
                alternateRefs: [],
            },
        ]);
    }));
    test('with custom transform', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { trailingSlash: true, transform: (_, url) => {
                if (!['/', '/page-2'].includes(url)) {
                    return;
                }
                return {
                    loc: url,
                    changefreq: 'yearly',
                };
            } }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'yearly',
                loc: 'https://example.com/',
                alternateRefs: [],
            },
            {
                changefreq: 'yearly',
                loc: 'https://example.com/page-2/',
                alternateRefs: [],
            },
        ]);
    }));
    test('with alternateRefs', () => __awaiter(void 0, void 0, void 0, function* () {
        const urlset = yield __1.createUrlSet(Object.assign(Object.assign({}, config_1.sampleConfig), { siteUrl: 'https://example.com/', alternateRefs: [
                { href: 'https://en.example.com/', hreflang: 'en' },
                { href: 'https://fr.example.com/', hreflang: 'fr' },
            ] }), manifest_1.sampleManifest);
        expect(urlset).toStrictEqual([
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com',
                alternateRefs: [
                    { href: 'https://en.example.com', hreflang: 'en' },
                    { href: 'https://fr.example.com', hreflang: 'fr' },
                ],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-0',
                alternateRefs: [
                    { href: 'https://en.example.com/page-0', hreflang: 'en' },
                    { href: 'https://fr.example.com/page-0', hreflang: 'fr' },
                ],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-1',
                alternateRefs: [
                    { href: 'https://en.example.com/page-1', hreflang: 'en' },
                    { href: 'https://fr.example.com/page-1', hreflang: 'fr' },
                ],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-2',
                alternateRefs: [
                    { href: 'https://en.example.com/page-2', hreflang: 'en' },
                    { href: 'https://fr.example.com/page-2', hreflang: 'fr' },
                ],
            },
            {
                changefreq: 'daily',
                lastmod: expect.any(String),
                priority: 0.7,
                loc: 'https://example.com/page-3',
                alternateRefs: [
                    { href: 'https://en.example.com/page-3', hreflang: 'en' },
                    { href: 'https://fr.example.com/page-3', hreflang: 'fr' },
                ],
            },
        ]);
    }));
});
